lib/phoenix_component.ex

```diff
@@ -1025,13 +1221,146 @@ defmodule Phoenix.Component do
+    slots
+  end
+
+  defp raise_if_function_already_defined!(env, name, slots, attrs) do
     if Module.defines?(env.module, {name, 1}) do
       {:v1, _, meta, _} = Module.get_definition(env.module, {name, 1})
-      [%{line: first_attr_line} | _] = attrs
 
-      compile_error!(first_attr_line, env.file, """
-      attributes must be defined before the first function clause at line #{meta[:line]}
+      unless Enum.empty?(attrs) do
+        [%{line: first_attr_line} | _] = attrs
+
+        compile_error!(first_attr_line, env.file, """
+        attributes must be defined before the first function clause at line #{meta[:line]}
+        """)
+      end
+
+      unless Enum.empty?(slots) do
+        [%{line: first_slot_line} | _] = slots
+
+        compile_error!(first_slot_line, env.file, """
+        slots must be defined before the first function clause at line #{meta[:line]}
+        """)
+      end
+    end
+  end
+
+  defmacro slot(name, opts \\ []) when is_atom(name) and is_list(opts) do
+    {block, opts} = Keyword.pop(opts, :do, nil)
+
+    quote do
+      Phoenix.Component.__slot__!(
+        __MODULE__,
+        unquote(name),
+        unquote(opts),
+        __ENV__.line,
+        __ENV__.file,
+        fn -> unquote(block) end
+      )
+    end
+  end
+
+  defmacro slot(name, opts, do: block) when is_atom(name) and is_list(opts) do
+    quote do
+      Phoenix.Component.__slot__!(
+        __MODULE__,
+        unquote(name),
+        unquote(opts),
+        __ENV__.line,
+        __ENV__.file,
+        fn -> unquote(block) end
+      )
+    end
+  end
+
+  @doc false
+  def __slot__!(module, name, opts, line, file, block_fun \\ fn -> :ok end) do
+    {doc, opts} = Keyword.pop(opts, :doc, nil)
+
+    unless is_binary(doc) or is_nil(doc) or doc == false do
+      compile_error!(line, file, ":doc must be a string or false, got: #{inspect(doc)}")
+    end
+
+    {required, opts} = Keyword.pop(opts, :required, false)
+
+    unless is_boolean(required) do
+      compile_error!(line, file, ":required must be a boolean, got: #{inspect(required)}")
+    end
+
+    Module.put_attribute(module, :__slot__, name)
+
+    try do
+      block_fun.()
+    rescue
+      e ->
+        Module.put_attribute(module, :__slot__, nil)
+        Module.delete_attribute(module, :__slot_attrs__)
+        Module.register_attribute(module, :__slot_attrs__, accumulate: true)
+        reraise e, __STACKTRACE__
+    end
+
+    slot_attrs =
+      module
+      |> Module.get_attribute(:__slot_attrs__)
+      |> Enum.reverse()
+
+    slot = %{
+      name: name,
+      required: required,
+      opts: opts,
+      doc: doc,
+      line: line,
+      attrs: slot_attrs
+    }
+
+    validate_slot!(module, slot, line, file)
+
+    Module.put_attribute(module, :__slots__, slot)
+    Module.put_attribute(module, :__slot__, nil)
+    Module.delete_attribute(module, :__slot_attrs__)
+    Module.register_attribute(module, :__slot_attrs__, accumulate: true)
```

Comment:
Based on the current implementation, I think we don't need reraise :)

```suggestion
    slot_attrs = 
      try do
        block_fun.()
        module |> Module.get_attribute(:__slot_attrs__) |> Enum.reverse()
      after
        Module.put_attribute(module, :__slot__, nil)
        Module.delete_attribute(module, :__slot_attrs__)
    end

    slot = %{
      name: name,
      required: required,
      opts: opts,
      doc: doc,
      line: line,
      attrs: slot_attrs
    }

    validate_slot!(module, slot, line, file)
    Module.put_attribute(module, :__slots__, slot)
    :ok
```